File: OpenRobertaServer/src/test/resources/crossCompilerTests/_expected/robotSpecific/targetLanguage/botnroll/math_lists.cpp
#include <Arduino.h>
#include <NEPODefs.h>
#include <BnrOneA.h>   // Bot'n Roll ONE A library 
#include <BnrRescue.h>   // Bot'n Roll CoSpace Rescue Module library 
#include <BnrRoberta.h>    // Open Roberta library 
BnrOneA one; 
BnrRescue brm; 
BnrRoberta bnr(one, brm);  
#define SSPIN  2 
#define MODULE_ADDRESS 0x2C 
byte colorsLeft[3]={0,0,0}; 
byte colorsRight[3]={0,0,0};
void ____math();
void ____lists();

inline bool _isPrime(double d);

double ___numberVar;
bool ___booleanVar;
String ___stringVar;
unsigned int ___colourVar;
std::list<double> ___numberList;
std::list<bool> ___booleanList;
std::list<String> ___stringList;
std::list<unsigned int> ___colourList;

void ____math() {
    one.lcd1(0);
    one.lcd1(___numberVar + ___numberVar);
    one.lcd1(___numberVar - ___numberVar);
    one.lcd1(___numberVar * ___numberVar);
    one.lcd1(___numberVar / ((float) ___numberVar));
    one.lcd1(pow(___numberVar, ___numberVar));
    one.lcd1(sqrt(___numberVar));
    one.lcd1(abs(___numberVar));
    one.lcd1(- (___numberVar));
    one.lcd1(log(___numberVar));
    one.lcd1(log10(___numberVar));
    one.lcd1(exp(___numberVar));
    one.lcd1(pow(10.0, ___numberVar));
    one.lcd1(sin(M_PI / 180.0 * (___numberVar)));
    one.lcd1(cos(M_PI / 180.0 * (___numberVar)));
    one.lcd1(tan(M_PI / 180.0 * (___numberVar)));
    one.lcd1(180.0 / M_PI * asin(___numberVar));
    one.lcd1(180.0 / M_PI * acos(___numberVar));
    one.lcd1(180.0 / M_PI * atan(___numberVar));
    one.lcd1(M_PI);
    one.lcd1(M_E);
    one.lcd1(M_GOLDEN_RATIO);
    one.lcd1(M_SQRT2);
    one.lcd1(M_SQRT1_2);
    one.lcd1(M_INFINITY);
    one.lcd1(bnr.boolToString((fmod(___numberVar, 2) == 0)));
    one.lcd1(bnr.boolToString((fmod(___numberVar, 2) != 0)));
    one.lcd1(bnr.boolToString(_isPrime(___numberVar)));
    one.lcd1(bnr.boolToString((___numberVar == floor(___numberVar))));
    one.lcd1(bnr.boolToString((___numberVar > 0)));
    one.lcd1(bnr.boolToString((___numberVar < 0)));
    one.lcd1(bnr.boolToString((fmod(___numberVar,___numberVar) == 0)));
    ___numberVar += ___numberVar;
    one.lcd1(round(___numberVar));
    one.lcd1(ceil(___numberVar));
    one.lcd1(floor(___numberVar));
    one.lcd1(_getListSum(___numberList));
    one.lcd1(_getListMin(___numberList));
    one.lcd1(_getListMax(___numberList));
    one.lcd1(_getListAverage(___numberList));
    one.lcd1(_getListMedian(___numberList));
    one.lcd1(_getListStandardDeviation(___numberList));
    one.lcd1(_getListElementByIndex(___numberList, 0));
    one.lcd1(fmod(___numberVar, ___numberVar));
    one.lcd1(std::min(std::max((double) ___numberVar, (double) ___numberVar), (double) ___numberVar));
    one.lcd1(_randomIntegerInRange(___numberVar, ___numberVar));
    one.lcd1(((double) rand() / (RAND_MAX)));
}

void ____lists() {
    ___numberList = {};
    ___numberList = {0, 0};
    ___numberList = _createListRepeat(___numberVar, (double) ___numberVar);
    ___booleanList = _createListRepeat(___numberVar, ___booleanVar);
    ___stringList = _createListRepeat(___numberVar, (String) ___stringVar);
    ___colourList = _createListRepeat(___numberVar, ___colourVar);
    one.lcd1(((int) ___numberList.size()));
    one.lcd1(___numberList.empty());
    one.lcd1(_getFirstOccuranceOfElement(___numberList, ___numberVar));
    one.lcd1(_getLastOccuranceOfElement(___numberList, ___numberVar));
    one.lcd1(_getListElementByIndex(___numberList, ___numberVar));
    one.lcd1(_getListElementByIndex(___numberList, ___numberList.size() - 1 - ___numberVar));
    one.lcd1(_getListElementByIndex(___numberList, 0));
    one.lcd1(_getListElementByIndex(___numberList, ___numberList.size() - 1));
    one.lcd1(_getAndRemoveListElementByIndex(___numberList, ___numberVar));
    one.lcd1(_getAndRemoveListElementByIndex(___numberList, ___numberList.size() - 1 - ___numberVar));
    one.lcd1(_getAndRemoveListElementByIndex(___numberList, 0));
    one.lcd1(_getAndRemoveListElementByIndex(___numberList, ___numberList.size() - 1));
    _removeListElementByIndex(___numberList, ___numberVar);
    _removeListElementByIndex(___numberList, ___numberList.size() - 1 - ___numberVar);
    _removeListElementByIndex(___numberList, 0);
    _removeListElementByIndex(___numberList, ___numberList.size() - 1);
}

void setup()
{
    Wire.begin();
    Serial.begin(9600);   // sets baud rate to 9600bps for printing values at serial monitor.
    one.spiConnect(SSPIN);   // starts the SPI communication module
    brm.i2cConnect(MODULE_ADDRESS);   // starts I2C communication
    brm.setModuleAddress(0x2C);
    one.stop();
    bnr.setOne(one);
    bnr.setBrm(brm);
    ___numberVar = 0;
    ___booleanVar = true;
    ___stringVar = "";
    ___colourVar = RGB(0xFF, 0xFF, 0xFF);
    ___numberList = {0, 0};
    ___booleanList = {true, true};
    ___stringList = {"", ""};
    ___colourList = {RGB(0xFF, 0xFF, 0xFF), RGB(0xFF, 0xFF, 0xFF)};

}

void loop()
{
    ____math();
    ____lists();
}
inline bool _isPrime(double d) {
    if (!(d == floor(d))) {
        return false;
    }
    int n = (int)d;
    if (n < 2) {
        return false;
    }
    if (n == 2) {
        return true;
    }
    if (n % 2 == 0) {
        return false;
    }
    for (int i = 3, s = (int)(sqrt(d) + 1); i <= s; i += 2) {
        if (n % i == 0) {
            return false;
        }
    }
    return true;
}

File: OpenRobertaServer/src/test/resources/crossCompilerTests/_expected/robotSpecific/targetLanguage/nao/math_lists.py
#!/usr/bin/python

import math
import time
import random
from roberta import Hal
h = Hal()


___numberVar = 0
___booleanVar = True
___stringVar = ""
___colourVar = 0xff0000
___numberList = [0, 0]
___booleanList = [True, True]
___stringList = ["", ""]
___colourList = [0xff0000, 0xff0000]
def ____math2():
    global ___numberVar, ___booleanVar, ___stringVar, ___colourVar, ___numberList, ___booleanList, ___stringList, ___colourList
    h.say(str(0))
    h.say(str(___numberVar + ___numberVar))
    h.say(str(___numberVar - ___numberVar))
    h.say(str(___numberVar * ___numberVar))
    h.say(str(___numberVar / float(___numberVar)))
    h.say(str(math.pow(___numberVar, ___numberVar)))
    h.say(str(math.sqrt(___numberVar)))
    h.say(str(math.fabs(___numberVar)))
    h.say(str(- (___numberVar)))
    h.say(str(math.log(___numberVar)))
    h.say(str(math.log10(___numberVar)))
    h.say(str(math.exp(___numberVar)))
    h.say(str(math.pow(10, ___numberVar)))
    h.say(str(math.sin(___numberVar)))
    h.say(str(math.cos(___numberVar)))
    h.say(str(math.tan(___numberVar)))
    h.say(str(math.asin(___numberVar)))
    h.say(str(math.acos(___numberVar)))
    h.say(str(math.atan(___numberVar)))
    h.say(str(math.pi))
    h.say(str(math.e))
    h.say(str((1 + 5 ** 0.5) / 2))
    h.say(str(math.sqrt(2)))
    h.say(str(math.sqrt(0.5)))
    h.say(str(float('inf')))
    h.say(str((___numberVar % 2) == 0))
    h.say(str((___numberVar % 2) == 1))
    h.say(str(_isPrime(___numberVar)))
    h.say(str((___numberVar % 1) == 0))
    h.say(str(___numberVar > 0))
    h.say(str(___numberVar < 0))
    h.say(str((___numberVar % ___numberVar) == 0))
    ___numberVar += ___numberVar
    h.say(str(round(___numberVar)))
    h.say(str(math.ceil(___numberVar)))
    h.say(str(math.floor(___numberVar)))
    h.say(str(sum(___numberList)))
    h.say(str(min(___numberList)))
    h.say(str(max(___numberList)))
    h.say(str(float(sum(___numberList))/len(___numberList)))
    h.say(str(_median(___numberList)))
    h.say(str(_standard_deviation(___numberList)))
    h.say(str(___numberList[0]))
    h.say(str(___numberVar % ___numberVar))
    h.say(str(min(max(___numberVar, ___numberVar), ___numberVar)))
    h.say(str(random.randint(___numberVar, ___numberVar)))
    h.say(str(random.random()))

def ____lists():
    global ___numberVar, ___booleanVar, ___stringVar, ___colourVar, ___numberList, ___booleanList, ___stringList, ___colourList
    ___numberList = []
    ___numberList = [0, 0]
    ___numberList = [___numberVar] * 5
    h.say(str(len( ___numberList)))
    h.say(str(not ___numberList))
    h.say(str(___numberList.index(___numberVar)))
    h.say(str((len(___numberList) - 1) - ___numberList[::-1].index(___numberVar)))
    h.say(str(___numberList[___numberVar]))
    h.say(str(___numberList[-1 -___numberVar]))
    h.say(str(___numberList[0]))
    h.say(str(___numberList[-1]))
    h.say(str(___numberList.pop(___numberVar)))
    h.say(str(___numberList.pop(-1 -___numberVar)))
    h.say(str(___numberList.pop(0)))
    h.say(str(___numberList.pop(-1)))
    ___numberList.pop(___numberVar)
    ___numberList.pop(-1 -___numberVar)
    ___numberList.pop(0)
    ___numberList.pop(-1)
    ___numberList[___numberVar] = ___numberVar
    ___numberList[-1 -___numberVar] = ___numberVar
    ___numberList[0] = ___numberVar
    ___numberList[-1] = ___numberVar
    ___numberList.insert(___numberVar, ___numberVar)
    ___numberList.insert(-1 -___numberVar, ___numberVar)
    ___numberList.insert(0, ___numberVar)
    ___numberList.insert(-1, ___numberVar)
    h.say(str(___numberList[___numberVar:___numberVar]))
    h.say(str(___numberList[___numberVar:-1 -___numberVar]))
    h.say(str(___numberList[___numberVar:]))
    h.say(str(___numberList[-1 -___numberVar:___numberVar]))
    h.say(str(___numberList[-1 -___numberVar:-1 -___numberVar]))
    h.say(str(___numberList[-1 -___numberVar:]))
    h.say(str(___numberList[0:___numberVar]))
    h.say(str(___numberList[0:-1 -___numberVar]))
    h.say(str(___numberList[0:]))


def run():
    h.setAutonomousLife('ON')
    global ___numberVar, ___booleanVar, ___stringVar, ___colourVar, ___numberList, ___booleanList, ___stringList, ___colourList
    ____math2()
    ____lists()

def main():
    try:
        run()
    except Exception as e:
        h.say("Error!" + str(e))
    finally:
        h.myBroker.shutdown()

def _median(l):
    l = sorted(l)
    l_len = len(l)
    if l_len < 1:
        return None
    if l_len % 2 == 0:
        return (l[int((l_len - 1) / 2)] + l[int((l_len + 1) / 2)] ) / 2.0
    else:
        return l[int((l_len - 1) / 2)]

def _isPrime(number):
    if(number == 0 or number == 1):
        return False
    for i in range(2, int(math.floor(math.sqrt(number))) + 1):
        remainder = number % i
        if remainder == 0:
            return False
    return True

def _standard_deviation(l):
    mean = float(sum(l)) / len(l)
    sd = 0
    for i in l:
        sd += (i - mean)*(i - mean)
    return math.sqrt(sd / len(l))
if __name__ == "__main__":
    main()